### PR TITLE
add Fusionlayer Infinity ansible module into ansible package

### PR DESCRIPTION
**Pull request by MeganLiu**
##### SUMMARY

add Fusionlayer company Infinity  IPAM (software) ansible module into ansible package

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

infinity
##### ANSIBLE VERSION

```
ansible --2.4
```

##### ADDITIONAL INFORMATION

I am submitting the first Pull request for Fusionlayer company Infinity software to support ansible usage.

```

```

### CODE REVIEW

**Comment by toshywoshy** [comment]
You import "requests" and "json" in the beginning of your module, the request module is not build in by default, and gives an error, use the "fetch_url" module from the builtin utils to avoid these problems, also I think there is also a builtin module for json, using the builtin means less maintenance for you
eg : from ansible.module_utils.urls import fetch_url

**Comment by MeganLiu** [comment]
@toshywoshy  thanks   for the suggestion. I will change it right now.

**Comment by MeganLiu** [comment]
Get error when trying to use ansible.module_utils.urls


    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 632, in _execute_module
    (module_style, shebang, module_data, module_path) = self._configure_module(module_name=module_name, module_args=module_args, task_vars=task_vars)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 157, in _configure_module
    task_vars=task_vars, module_compression=self._play_context.module_compression)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 796, in modify_module
    (b_module_data, module_style, shebang) = _find_module_utils(module_name, b_module_data, module_path, module_args, task_vars, module_compression)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 678, in _find_module_utils
    recursive_finder(module_name, b_module_data, py_module_names, py_module_cache, zf)
  File "/usr/lib/python2.7/site-packages/ansible/executor/module_common.py", line 504, in recursive_finder
    msg.append('either %s.py or %s.py' % (py_module_name[-1], py_module_name[-2]))
IndexError: tuple index out of range

**Comment by alikins** [comment]
The module is in the wrong path (lib/ansible/module_utils/infinity/infinity.py). It needs to be moved to the modules path (lib/ansible/modules/*/infinity.py). I'm not sure what category this fits in (the '*') but whatever is approriate.

**Review by alikins — CHANGES_REQUESTED** [review]
The infinity.py module is in the wrong path. 

It needs to be moved to lib/ansible/modules/ instead of lib/ansible/module_utils/

**Comment by MeganLiu** [comment]
Thanks,I will change the path of the file.:)
Meirong

2017年7月17日 18:56，"Adrian Likins" <notifications@github.com>写道：

**Comment by MeganLiu** [comment]
Yes,it's about IPAddress management.Do  you mean this category?

2017年7月17日 18:56，"Adrian Likins" <notifications@github.com>写道：

**Comment by MeganLiu** [comment]
Based on the review suggestions, I tried to use ansible built-in module "uri" to send rest api request. However, the "url" "post" method could not send the data to the server side.
Is it okay that I use python library " urllib2 "  to send REST API call?
Looking forward to your suggestion. Thanks.

**Comment by MeganLiu** [comment]
hi, 
I have made code changes to remove dependency on requests module. Could you please let me know  what do I need to do next?

**Comment by MeganLiu** [comment]
HI
I tried to use   ansible.url to post data but it did not work . So I use python  urllib2 to post data.
In this situation, what should I do ? report a bug to url and wait for reply ?

**Comment by MeganLiu** [comment]
Hi, 
what needs to be done next ?

**Comment by MeganLiu** [comment]
Thanks for the feedback. Now the file is in the path  "lib/ansible/modules/network/infinity"

**Comment by MeganLiu** [comment]
Hi, Alikins

for the shippable testing, 30488 failure, could you please let me know what should I do?  Thanks.

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: This might be breaking the yaml parser, try rewriting as

```
choices:
  - reserve_next_available_ip
  - release_ip

```

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: `"2.4"`

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: Don't import `*`, import only what you need

See: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#python-imports

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: `Manage Infinity IPAM using  Web API`

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: or hostname?

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: `- Username to access Infinity.`

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: "indicated" how/

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: `required: False` lines can be removed as that's the default.

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: Have a look at the `RETURN` block in another module along side the documentation at http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: You need `no_log='True'` here as well

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: This seems a lot longer and less readable that the standard way of defining the `argument_spec`.
https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: Is this needed?

If they are then they should be added to the `DOCUMENTATION`

For usage see https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec

http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: `result2` doesn't seem like a very clear variable name. It's difficult to follow what's going on here

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: Shouldn't this be a failure, rather than just `exit_json`?

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: Does this not work?
If so delete it or fix it in this PR

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: `#!/usr/bin/python`

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: All modules must have the following sections defined in this order:

All modules must have the following sections defined in this order:

Copyright
ANSIBLE_METADATA
DOCUMENTATION
EXAMPLES
RETURNS
Python imports


See http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html

File: lib/ansible/modules/network/infinity/infinity.py
Comment (gundalow) [inline lib/ansible/modules/network/infinity/infinity.py]: Need to move *after* `RETURN`

**Review by gundalow — CHANGES_REQUESTED** [review]
You can test this locally with

`./test/runner/ansible-test sanity --test validate-modules infinity` So you can identify and test your fixes locally without having to wait on Shippable.

I believe if you read http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html that will answer most of your questions.

**Comment by MeganLiu** [comment]
Hi, John

Many thanks for the help :). I will make the change.
Meirong

**Comment by MeganLiu** [comment]
Thanks for the feedback :). I am making the changes.

**Comment by MeganLiu** [comment]
Could someone provide the suggestions on what should I do the the failure?

**Comment by MeganLiu** [comment]
Could you please let me know what needs to be done next when you see my message ?   Thanks.

**Comment by MeganLiu** [comment]
Could you please let me know what needs to be done next  since the module has passed the test
  ?

**Comment by MeganLiu** [comment]
hi,  
could you kindly let me know whether I  need to do something else  before merge the code into ansible repo?

**Comment by gundalow** [comment]
Code looks good, just one minor change, can you please move the files to `modules/net_tools` (rather than `modules/network` thanks.

**Comment by MeganLiu** [comment]
Thanks for providing the new comments :)  I will change the directory right now

**Comment by MeganLiu** [comment]
@ansibot  and @gundalow : could you please let me know whether I need to do anything more  before this module is merged into ansible repo ? I would like this module to be shipped with ansible 2.4 release.  I really appreciate your further review comments (if there are) .

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged, will be released in Ansible 2.4.0

**Pull request merged**
